### PR TITLE
feat(locking): adds locking to traffic guarded operations

### PR DESCRIPTION
**Pull request by cfieber**
Uses TrafficGuard as a signal that a lock should be held on a cluster
for an operation. Adds locking to
* deploy stages via AbstractDeployStrategyStage
* TargetServerGroup stages via TargetServerGroupLinearStageSupport
* Cluster-wide stages via AbstractClusterWideClouddriverOperationStage

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/AbstractClusterWideClouddriverTask.groovy [hunk 1]
```diff
@@ -111,18 +94,7 @@ abstract class AbstractClusterWideClouddriverTask extends AbstractCloudProviderA
...
-      : stage.context.regions
-        ? stage.context.regions.collect { new Location(type: Location.Type.REGION, value: it) }
-        : stage.context.zones
-          ? stage.context.zones.collect { new Location(type: Location.Type.ZONE, value: it) }
-          : stage.context.namespace
-            ? [new Location(type: Location.Type.NAMESPACE, value: stage.context.namespace)]
-            : stage.context.region
-              ? [new Location(type: Location.Type.REGION, value: stage.context.region)]
-              : []
+    def locations = AbstractClusterWideClouddriverOperationStage.locationsFromStage(stage.context)
```
Comment (robzienert) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/AbstractClusterWideClouddriverTask.groovy]: :105:

**Comment by cfieber** [comment]
@ezimanyi ahh, great. It also breaks custom strategies so I'm working on a few fixes around this one currently

**Comment by ezimanyi** [comment]
Whoops, sorry for tagging Adam by mistake---was discussing with him on another task immediately before this and git mixed up.

**Comment by cfieber** [comment]
I have a fix locally for the application auth (and a couple other issues this change introduced), will get a PR up asap

**Pull request merged**